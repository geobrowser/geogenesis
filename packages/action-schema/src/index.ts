import Ajv from 'ajv'
import schema from '../build/schema.json'
import { Root } from './action'

const ajv = new Ajv()

export * from './action'

export const ActionSchema = {
  schema,
  validate: ajv.compile<Root>(schema),
}
