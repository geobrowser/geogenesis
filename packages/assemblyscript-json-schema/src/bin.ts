#!/usr/bin/env node

let { cli } = require('./cli')

cli({ argv: process.argv.slice(2) })
