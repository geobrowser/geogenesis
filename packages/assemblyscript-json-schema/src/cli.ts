import arg from 'arg'
import { generate } from './index'

export async function cli({ argv }: { argv: string[] }) {
  const result = arg(
    {
      '--help': Boolean,
      '--schema': String,
      '--output': String,
      '--static': String,
    },
    { argv }
  )

  const { '--schema': schema, '--output': output, '--static': static_ } = result

  if (result['--help']) {
    console.log(
      `assemblyscript-json-schema --schema <SCHEMA_PATH> --static <STATIC_PATH> --output <OUTPUT_PATH>`
    )
    return process.exit(0)
  }

  if (!schema) {
    console.error(`Missing 'schema' arg`)
    return process.exit(1)
  }
  if (!output) {
    console.error(`Missing 'output' arg`)
    return process.exit(1)
  }

  generate({
    paths: {
      schemaPath: schema,
      outputPath: output,
      staticPath: static_,
    },
  })
}
