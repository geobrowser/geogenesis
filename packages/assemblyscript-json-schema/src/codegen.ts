import { JSONSchema7 } from 'json-schema'

import {
  findRefs,
  getAnyOfTypeNames,
  getDefinition,
  getDiscriminator,
  getRefName,
  getUnionsContainingType,
  hasArrayOfComplexTypes,
  unique,
} from './schemaUtils'

function generateImport(name: string) {
  return `import { ${name} } from "./${name}"`
}

export function generateUnionType(
  schema: JSONSchema7,
  name: string,
  definition: JSONSchema7
) {
  const anyOfTypeNames = getAnyOfTypeNames(definition)

  return `
  import { log } from '@graphprotocol/graph-ts'
  import { JSON } from 'assemblyscript-json/assembly'
  ${anyOfTypeNames.map(generateImport).join('\n')}

  export class ${name} {
    type: string

    constructor(type: string) {
      this.type = type;
    }

    ${anyOfTypeNames
      .map(
        (typeName) =>
          `as${typeName}(): ${typeName} | null {
            return this.type == '${getDiscriminator(
              getDefinition(schema, typeName)
            )}' ? (this as ${typeName}) : null;
          }`
      )
      .join('\n\n')}

    toJSON(): JSON.Value {
      ${anyOfTypeNames
        .map(
          (typeName) =>
            `if (this.type == '${getDiscriminator(
              getDefinition(schema, typeName)
            )}') return (this as ${typeName}).toJSON()`
        )
        .join('\n')}
      throw \`undefined variant of: ${name}.\${this.type}\`
    }

    static fromJSON(__json: JSON.Value): ${name} | null {
      if (!__json.isObj) {
        log.debug("${name}.fromJSON(): __json.isObj is false", [])
        return null
      }
      const __obj = __json as JSON.Obj
      const type = __obj.getString('type')
      if (type == null) {
        log.debug("${name}.fromJSON(): type is null", [])
        return null
      }
      const typeName = type.valueOf()
      ${anyOfTypeNames
        .map(
          (typeName) =>
            `if (typeName == '${getDiscriminator(
              getDefinition(schema, typeName)
            )}') return ${typeName}.fromJSON(__json)`
        )
        .join('\n')}
      log.debug(\`${name}.fromJSON(): unhandled variant '\${typeName}'\`, [])
      return null;
    }
  }
  `
}

export function generateObjectType(
  schema: JSONSchema7,
  name: string,
  definition: JSONSchema7
) {
  const { properties = {} } = definition
  const unions = getUnionsContainingType(schema, name)
  const importedNames = unique([
    ...findRefs(definition).map(getRefName),
    ...unions,
  ])

  return `
  import { log } from '@graphprotocol/graph-ts'
  import { JSON } from 'assemblyscript-json/assembly'
  ${importedNames.map(generateImport).join('\n')}
  ${
    hasArrayOfComplexTypes(definition)
      ? `import { mapOrNull } from "./collection-utils"`
      : ''
  }  

  export class ${name} ${
    unions.length > 0 ? `extends ${unions.join(', ')}` : ''
  } {
    ${Object.entries(properties)
      .filter(([property]) => unions.length === 0 || property !== 'type')
      .map(
        ([property, value]) =>
          `${property}: ${convertTypeName(schema, value as JSONSchema7)}`
      )
      .join('\n')}
      
    constructor(${Object.entries(properties).map(
      ([property, value]) =>
        `${property}: ${convertTypeName(schema, value as JSONSchema7)}`
    )}) {
      ${unions.length > 0 ? 'super(type)' : ''}
      ${Object.keys(properties)
        .map((property) => `this.${property} = ${property}`)
        .join('\n')}
    }

    toJSON(): JSON.Value {
      const __obj = new JSON.Obj()
      ${Object.entries(properties)
        .map(([property, value]) => {
          let result = ``

          if ((value as JSONSchema7).type === 'array') {
            result += `const ${property} = new JSON.Arr()
            for (let i = 0; i < this.${property}.length; i++) {
              ${property}.push(this.${property}[i].toJSON())
            }`
          } else if ((value as JSONSchema7).$ref) {
            result += `const ${property} = this.${property}.toJSON()`
          } else {
            result += `const ${property} = ${getEncoderFunction(
              value as JSONSchema7
            )}(this.${property})`
          }

          result += `
          __obj.set('${property}', ${property})
          `

          return result.trim()
        })
        .join('\n')}

      return __obj
    }

    static fromJSON(__json: JSON.Value): ${name} | null {
      if (!__json.isObj) {
        log.debug("${name}.fromJSON(): __json.isObj is false", [])
        return null
      }
      const __obj = __json as JSON.Obj
      ${Object.entries(properties)
        .map(([property, value]) => {
          const valueSchemaType = (value as JSONSchema7).type
          let result = `
          let __${property} = ${getDecoderFunction(
            property,
            value as JSONSchema7
          )}
          ${
            valueSchemaType === 'number'
              ? `if (__${property} == null) {
                  const integer = __obj.getInteger("${property}")
                  if (integer != null) {
                    __${property} = new JSON.Num(integer.valueOf() as f64)
                  }
                }`
              : ''
          }
          if (__${property} == null) {
            log.debug("${name}.fromJSON(): __${property} is null", [])
            return null
          }
          `
          if (valueSchemaType === 'array') {
            const itemType = (value as JSONSchema7).items! as JSONSchema7
            const refName = getRefName(itemType.$ref!)
            result += `const __${property}Array = __${property}.valueOf()
            const ${property} = mapOrNull<JSON.Value, ${refName}>(
              __${property}Array,
              (item: JSON.Value): ${refName} | null => ${refName}.fromJSON(item)
            )
            if (${property} == null) {
              log.debug("${name}.fromJSON(): __${property} is null", [])
              return null
            }`
          } else if ((value as JSONSchema7).$ref) {
            const refName = getRefName((value as JSONSchema7).$ref!)
            result += `const ${property} = ${refName}.fromJSON(__${property})
            if (${property} == null) {
              log.debug("${name}.fromJSON(): __${property} is null", [])
              return null
            }`
          } else {
            result += `const ${property} = __${property}.valueOf()`
          }

          return result.trim()
        })
        .join('\n')}

      return new ${name}(
        ${Object.keys(properties)}
      )
    }
  }
  `
}

export function generateType(schema: JSONSchema7, name: string) {
  const definition = getDefinition(schema, name)

  if (definition.properties) {
    return generateObjectType(schema, name, definition)
  }

  if (definition.anyOf) {
    return generateUnionType(schema, name, definition)
  }

  throw new Error(`Unrecognized type: ${name}`)
}

function convertTypeName(schema: JSONSchema7, property: JSONSchema7) {
  switch (property.type) {
    case 'boolean':
    case 'string':
      return property.type
    case 'number':
      return 'f64'
    case 'array':
      const name = getRefName((property.items as JSONSchema7).$ref!)
      return `${name}[]`
  }

  if (property.$ref) {
    return getRefName(property.$ref)
  }

  throw new Error(`Failed to get type name: ${JSON.stringify(property)}`)
}

function getDecoderFunction(property: string, value: JSONSchema7) {
  if (typeof value.type === 'string') {
    switch (value.type) {
      case 'number':
        return `__obj.getNum("${property}")`
      case 'string':
        return `__obj.getString("${property}")`
      case 'boolean':
        return `__obj.getBool("${property}")`
      case 'array':
        return `__obj.getArr("${property}")`
      default:
        break
    }
  }

  if (typeof value.$ref === 'string') {
    return `__obj.getObj("${property}")`
  }

  throw new Error(`Unsupported property: ${JSON.stringify(value)}`)
}

function getEncoderFunction(value: JSONSchema7) {
  if (typeof value.type === 'string') {
    switch (value.type) {
      case 'number':
        return `new JSON.Num`
      case 'string':
        return `new JSON.Str`
      case 'boolean':
        return `new JSON.Bool`
      case 'array':
        return `new JSON.Arr`
      default:
        break
    }
  }

  if (typeof value.$ref === 'string') {
    return `new JSON.Obj`
  }

  throw new Error(`Unsupported property: ${JSON.stringify(value)}`)
}
