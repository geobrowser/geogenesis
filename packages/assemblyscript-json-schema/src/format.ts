import { IFS } from 'buffs'
import path from 'path'
import prettier from 'prettier'

export function formatTypeScript(source: string) {
  return prettier.format(source, {
    parser: 'typescript',
    singleQuote: true,
    trailingComma: 'es5',
    semi: false,
  })
}

/**
 * Format all files in a directory
 */
export async function formatDirectory(fs: IFS, directoryPath: string) {
  const files = await fs.promises.readdir(directoryPath)

  await Promise.all(
    files.map(async (name) => {
      if (path.extname(name) !== '.ts') return

      const filePath = path.join(directoryPath, name)
      const source = await fs.promises.readFile(filePath, 'utf8')
      const formatted = formatTypeScript(source)

      await fs.promises.writeFile(filePath, formatted)
    })
  )
}
