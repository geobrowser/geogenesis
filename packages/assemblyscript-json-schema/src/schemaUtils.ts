import { JSONSchema7 } from 'json-schema'
import get from 'lodash.get'
import { withOptions } from 'tree-visit'

type JSONValue =
  | null
  | boolean
  | number
  | string
  | { [x: string]: JSONValue }
  | Array<JSONValue>

const traverse = withOptions<[string, JSONValue]>({
  getChildren: ([, value]) => {
    if (typeof value === 'object' && value !== null) {
      return Object.entries(value)
    } else {
      return []
    }
  },
})

export function getTypeNames(schema: JSONSchema7) {
  return Object.keys(schema.definitions ?? {})
}

export function getDefinition(schema: JSONSchema7, name: string): JSONSchema7 {
  const definition = (schema.definitions ?? {})[name]

  if (typeof definition === 'boolean') {
    throw new Error(`Unhandled boolean definition for: ${name}`)
  }

  return definition
}

export function getResolvedReference<T = unknown>(
  schema: JSONSchema7,
  ref: string
): T {
  if (ref.startsWith('#')) {
    const components = ref.split('/').slice(1)
    return get(schema, components)
  }

  throw new Error(`Unable to resolve reference: ${ref}`)
}

export function getRefName($ref: string) {
  return $ref!.split('/').at(-1)!
}

export function getAnyOfTypeNames(definition: JSONSchema7) {
  const anyOf = (definition.anyOf || []).flatMap((schema) =>
    typeof schema === 'boolean' ? [] : [schema]
  )

  if (anyOf.some((option) => !option.$ref)) {
    throw new Error(`Missing anyOf ref in: ${name}`)
  }

  const typeNames = anyOf.map((option) => getRefName(option.$ref!))

  return typeNames
}

export function getUnionsContainingType(schema: JSONSchema7, name: string) {
  const typeNames = getTypeNames(schema)

  const unionTypeNames = typeNames.flatMap((typeName) => {
    const definition = getDefinition(schema, typeName)
    const anyOfTypeNames = getAnyOfTypeNames(definition)

    return anyOfTypeNames.includes(name) ? [typeName] : []
  })

  return unique(unionTypeNames)
}

export function unique<T>(array: T[]): T[] {
  return [...new Set(array)]
}

export function findRefs(schema: JSONSchema7) {
  return unique(
    traverse
      .findAll(['', schema as JSONValue], ([key]) => key === '$ref')
      .flatMap(([, value]) => (typeof value === 'string' ? [value] : []))
  )
}

export function hasArrayOfComplexTypes(schema: JSONSchema7) {
  return !!traverse.find(
    ['', schema as JSONValue],
    ([key, value]) =>
      key === 'items' &&
      typeof value === 'object' &&
      value != null &&
      '$ref' in value
  )
}

export function getDiscriminator(definition: JSONSchema7) {
  const properties = definition.properties || {}
  const typeProperty = properties.type as JSONSchema7 | undefined
  return typeof typeProperty?.const === 'string'
    ? typeProperty.const
    : undefined
}
