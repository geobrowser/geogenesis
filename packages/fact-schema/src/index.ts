import Ajv from 'ajv'
import schema from '../build/schema.json'
import { Root } from '../src/fact'

const ajv = new Ajv()

export * from './fact'

export const FactSchema = {
  schema,
  validate: ajv.compile<Root>(schema),
}
