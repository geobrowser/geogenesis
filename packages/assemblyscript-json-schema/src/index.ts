import { copy, createFs, IFS } from 'buffs'
import { JSONSchema7 } from 'json-schema'
import path from 'path'
import { generateType } from './codegen'
import { formatDirectory } from './format'

export interface Paths {
  schemaPath: string
  outputPath: string
  staticPath?: string
}

type Configuration = {
  paths: Paths
  inputFs?: IFS
  outputFs?: IFS
}

export async function generate({
  paths,
  inputFs = require('fs'),
  outputFs = require('fs'),
}: Configuration) {
  const schemaString = await inputFs.promises.readFile(paths.schemaPath, 'utf8')
  const schema = JSON.parse(schemaString) as JSONSchema7

  const generatedTypes = getTypeNames(schema).map((name) => [
    `/${name}.ts`,
    generateType(schema, name),
  ])

  const generatedFs = createFs(Object.fromEntries(generatedTypes))

  const staticPath = paths.staticPath ?? path.join(__dirname, '../static')

  // Copy all static files
  copy(inputFs, generatedFs, staticPath, '/')

  await generateIndexFile(generatedFs, '/')
  await formatDirectory(generatedFs, '/')

  // Write the generated files to the output fs
  copy(generatedFs, outputFs, '/', paths.outputPath)
}

export function getTypeNames(schema: JSONSchema7) {
  return Object.keys(schema.definitions ?? {})
}

/**
 * Generate an index file that re-exports all the source files in a directory
 */
async function generateIndexFile(fs: IFS, directoryPath: string) {
  const files = await fs.promises.readdir(directoryPath, 'utf8')

  const source = files
    .map((name) => `export * from "./${path.basename(name, '.ts')}"`)
    .join('\n')

  await fs.promises.writeFile(path.join(directoryPath, 'index.ts'), source)
}
