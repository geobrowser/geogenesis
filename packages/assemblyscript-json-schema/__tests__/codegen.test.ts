import { JSONSchema7 } from 'json-schema'
import { expect, it } from 'vitest'
import { generateType } from '../src/codegen'
import { formatTypeScript } from '../src/format'
import { getUnionsContainingType } from '../src/schemaUtils'

const schema: JSONSchema7 = {
  $ref: '#/definitions/Root',
  $schema: 'http://json-schema.org/draft-07/schema#',
  definitions: {
    NumberValue: {
      properties: {
        type: {
          const: 'number',
          type: 'string',
        },
        value: {
          type: 'number',
        },
      },
      required: ['type', 'value'],
      type: 'object',
    },
    StringValue: {
      properties: {
        type: {
          const: 'string',
          type: 'string',
        },
        value: {
          type: 'string',
        },
      },
      required: ['type', 'value'],
      type: 'object',
    },
    Value: {
      anyOf: [
        {
          $ref: '#/definitions/NumberValue',
        },
        {
          $ref: '#/definitions/StringValue',
        },
      ],
    },
    ValuesContainer: {
      properties: {
        values: {
          items: {
            $ref: '#/definitions/Value',
          },
          type: 'array',
        },
        type: {
          const: 'ValuesContainer',
          type: 'string',
        },
      },
      required: ['type', 'values'],
      type: 'object',
    },
    ValueContainer: {
      properties: {
        value: {
          $ref: '#/definitions/Value',
        },
        type: {
          const: 'ValueContainer',
          type: 'string',
        },
      },
      required: ['type', 'value'],
      type: 'object',
    },
  },
}

it('generates object type', async () => {
  const types = formatTypeScript(generateType(schema, 'NumberValue'))

  expect(types).toMatchSnapshot()
})

it('generates union type', async () => {
  const containing = getUnionsContainingType(schema, 'NumberValue')

  expect(containing).toEqual(['Value'])

  const types = formatTypeScript(generateType(schema, 'Value'))

  expect(types).toMatchSnapshot()
})

it('generates array type', async () => {
  const types = formatTypeScript(generateType(schema, 'ValuesContainer'))

  expect(types).toMatchSnapshot()
})

it('generates object type', async () => {
  const types = formatTypeScript(generateType(schema, 'ValueContainer'))

  expect(types).toMatchSnapshot()
})
