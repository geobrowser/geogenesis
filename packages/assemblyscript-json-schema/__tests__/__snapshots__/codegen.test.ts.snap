// Vitest Snapshot v1

exports[`generates array type 1`] = `
"import { log } from '@graphprotocol/graph-ts'
import { JSON } from 'assemblyscript-json/assembly'
import { Value } from './Value'
import { mapOrNull } from './collection-utils'

export class ValuesContainer {
  values: Value[]
  type: string

  constructor(values: Value[], type: string) {
    this.values = values
    this.type = type
  }

  toJSON(): JSON.Value {
    const __obj = new JSON.Obj()
    const values = new JSON.Arr()
    for (let i = 0; i < this.values.length; i++) {
      values.push(this.values[i].toJSON())
    }
    __obj.set('values', values)
    const type = new JSON.Str(this.type)
    __obj.set('type', type)

    return __obj
  }

  static fromJSON(__json: JSON.Value): ValuesContainer | null {
    if (!__json.isObj) {
      log.debug('ValuesContainer.fromJSON(): __json.isObj is false', [])
      return null
    }
    const __obj = __json as JSON.Obj
    let __values = __obj.getArr('values')

    if (__values == null) {
      log.debug('ValuesContainer.fromJSON(): __values is null', [])
      return null
    }
    const __valuesArray = __values.valueOf()
    const values = mapOrNull<JSON.Value, Value>(
      __valuesArray,
      (item: JSON.Value): Value | null => Value.fromJSON(item)
    )
    if (values == null) {
      log.debug('ValuesContainer.fromJSON(): __values is null', [])
      return null
    }
    let __type = __obj.getString('type')

    if (__type == null) {
      log.debug('ValuesContainer.fromJSON(): __type is null', [])
      return null
    }
    const type = __type.valueOf()

    return new ValuesContainer(values, type)
  }
}
"
`;

exports[`generates object type 1`] = `
"import { log } from '@graphprotocol/graph-ts'
import { JSON } from 'assemblyscript-json/assembly'
import { Value } from './Value'

export class NumberValue extends Value {
  value: f64

  constructor(value: f64) {
    super('number')
    this.value = value
  }

  toJSON(): JSON.Value {
    const __obj = new JSON.Obj()
    const type = new JSON.Str(this.type)
    __obj.set('type', type)
    const value = new JSON.Num(this.value)
    __obj.set('value', value)

    return __obj
  }

  static fromJSON(__json: JSON.Value): NumberValue | null {
    if (!__json.isObj) {
      log.debug('NumberValue.fromJSON(): __json.isObj is false', [])
      return null
    }
    const __obj = __json as JSON.Obj
    let __type = __obj.getString('type')

    if (__type == null) {
      log.debug('NumberValue.fromJSON(): __type is null', [])
      return null
    }
    const type = __type.valueOf()
    let __value = __obj.getNum('value')
    if (__value == null) {
      const integer = __obj.getInteger('value')
      if (integer != null) {
        __value = new JSON.Num(integer.valueOf() as f64)
      }
    }
    if (__value == null) {
      log.debug('NumberValue.fromJSON(): __value is null', [])
      return null
    }
    const value = __value.valueOf()

    return new NumberValue(value)
  }
}
"
`;

exports[`generates object type 2`] = `
"import { log } from '@graphprotocol/graph-ts'
import { JSON } from 'assemblyscript-json/assembly'
import { Value } from './Value'

export class ValueContainer {
  value: Value
  type: string

  constructor(value: Value, type: string) {
    this.value = value
    this.type = type
  }

  toJSON(): JSON.Value {
    const __obj = new JSON.Obj()
    const value = this.value.toJSON()
    __obj.set('value', value)
    const type = new JSON.Str(this.type)
    __obj.set('type', type)

    return __obj
  }

  static fromJSON(__json: JSON.Value): ValueContainer | null {
    if (!__json.isObj) {
      log.debug('ValueContainer.fromJSON(): __json.isObj is false', [])
      return null
    }
    const __obj = __json as JSON.Obj
    let __value = __obj.getObj('value')

    if (__value == null) {
      log.debug('ValueContainer.fromJSON(): __value is null', [])
      return null
    }
    const value = Value.fromJSON(__value)
    if (value == null) {
      log.debug('ValueContainer.fromJSON(): __value is null', [])
      return null
    }
    let __type = __obj.getString('type')

    if (__type == null) {
      log.debug('ValueContainer.fromJSON(): __type is null', [])
      return null
    }
    const type = __type.valueOf()

    return new ValueContainer(value, type)
  }
}
"
`;

exports[`generates union type 1`] = `
"import { log } from '@graphprotocol/graph-ts'
import { JSON } from 'assemblyscript-json/assembly'
import { NumberValue } from './NumberValue'
import { StringValue } from './StringValue'

export class Value {
  type: string

  constructor(type: string) {
    this.type = type
  }

  asNumberValue(): NumberValue | null {
    return this.type == 'number' ? (this as NumberValue) : null
  }

  asStringValue(): StringValue | null {
    return this.type == 'string' ? (this as StringValue) : null
  }

  toJSON(): JSON.Value {
    if (this.type == 'number') return (this as NumberValue).toJSON()
    if (this.type == 'string') return (this as StringValue).toJSON()
    throw \`undefined variant of: Value.\${this.type}\`
  }

  static fromJSON(__json: JSON.Value): Value | null {
    if (!__json.isObj) {
      log.debug('Value.fromJSON(): __json.isObj is false', [])
      return null
    }
    const __obj = __json as JSON.Obj
    const type = __obj.getString('type')
    if (type == null) {
      log.debug('Value.fromJSON(): type is null', [])
      return null
    }
    const typeName = type.valueOf()
    if (typeName == 'number') return NumberValue.fromJSON(__json)
    if (typeName == 'string') return StringValue.fromJSON(__json)
    log.debug(\`Value.fromJSON(): unhandled variant '\${typeName}'\`, [])
    return null
  }
}
"
`;
